platform/engineering/design-docs/2020-05-26-content-validation.md

```diff
@@ -0,0 +1,201 @@
+      - [Risks](#risks)
+      - [Questions](#questions)
+    - [Work Estimates](#work-estimates)
+    - [Alternatives](#alternatives)
+      - [Highest quality, but lowest speed](#highest-quality-but-lowest-speed)
+      - [Fastest speed, but lowest quality](#fastest-speed-but-lowest-quality)
+    - [Future Work](#future-work)
+    - [Revision History](#revision-history)
+
+## Overview
+
+### Objective
+
+The objective here is to define an updated approach to validate the CMS content. Specifically, this design doc focuses on validating that CMS content has no accessibility errors and/or broken links.
+
+This design document is intended for front end and DevOps engineers on the Veteran-facing Services Platform (VSP) and CMS teams.
+
+### Background
+
+The `vets-website` repo currently contains [one script that builds both the content and the applications](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/index.js). Chris V. wrote a [design doc to disentangle the content build from the application build](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/2020-04-09-separate-content-build.md), and has since broken the build into two two separate stages.
+
+1. [Run Webpack](https://github.com/department-of-veterans-affairs/vets-website/blob/43ea0bdccd5e53886e3e38ea27d3f8e8e7bd9038/script/build.sh#L55)
+2. [Run the content build](https://github.com/department-of-veterans-affairs/vets-website/blob/43ea0bdccd5e53886e3e38ea27d3f8e8e7bd9038/script/build.sh#L61)
+
+Previously, Webpack was entangled in the content build, but no longer. Brooks is currently working on moving the content build into a [`content-build`](https://github.com/department-of-veterans-affairs/content-build/) repo to separate it from `vets-website`.
+
+##### Content Sources
+
+Static content currently comes from two sources:
+
+1. The Drupal CMS
+2. The [`vagov-content` repo](https://github.com/department-of-veterans-affairs/vagov-content/blob/master/pages/404.md)
+
+This design doc will focus on validating content from the Drupal CMS.
+
+##### Amazon S3
+
+The frontend of VA.gov is served as static files via Amazon S3. Since `vets-website` currently handles both application and content building, a failure in one blocks both from being deployed to S3 during a full deploy.
+
+###### Broken Link Checking
+
+The [`check-broken-link` middleware](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/plugins/check-broken-links/index.js) is a Metalsmith plugin that happens near the end of the Metalsmith static content build pipeline. It does the following:
+
+1. Loop through all of the HTML files in the Metalsmith pipeline
+2. Extract all of the broken `href`/`src` values using the `getBrokenLinks` helper
+3. Formats the result into useful console output
+4. Blocks the deployment on production or logs the error output on lower environments
```

Comment:
Some historical info might be useful - originally (on vets.gov), all content was in vets-website in a `content/` directory and you would open a PR to edit content. CI would then run against that PR. Link-checking would run as part of the build the same way it does today. If a broken link was found, CI would fail, and you wouldn't be able to merge. This meant that broken links never made their way into the master branch.

Accessibility checks had the same deal - errors never made their way into master.

Also - I think Drupal contains some link-checking internally nowadays but it isn't as comprehensive or something.